src/ui/platforms/desktop/src/pages/room/actions/select_action.ts

```diff
@@ -87,6 +90,73 @@ export class SelectAction<T extends GameEventIdentifiers> extends BaseAction {
     });
   }
 
+  onSelectCardTargets(cardId: CardId, exclude: PlayerId[]) {
+    return new Promise<PlayerId[] | undefined>(resolve => {
+      this.presenter.delightPlayers(true);
+
+      const card = Sanguosha.getCardById(cardId);
+      if (!(card.Skill instanceof ActiveSkill)) {
+        resolve();
+      }
+
+      this.presenter.enableActionButton('cancel');
+      this.presenter.defineCancelButtonActions(() => {
+        this.resetAction();
+        resolve();
+      });
+
+      const selectedTargets: PlayerId[] = [];
+      const cardSkill = card.Skill as ActiveSkill;
+      const skillExtralProp = (card.Skill as unknown) as ExtralCardSkillProperty;
+      let numberOfCardTargets = cardSkill.numberOfTargets();
+      numberOfCardTargets = (numberOfCardTargets instanceof Array ? numberOfCardTargets[0] : numberOfCardTargets) || 1;
+
+      const excludeExistedTargets = (playerId: PlayerId) =>
+        (selectedTargets.length === 0 ? !exclude.includes(playerId) : true) &&
+        this.store.room.isAvailableTarget(cardId, this.store.clientPlayerId, playerId);
+      const isAvailableTarget = skillExtralProp.isCardAvailableTarget || cardSkill.isAvailableCard;
```

Comment:
it has a potential problem since you changed the `this` pointer from skill class instance into the action handlers, once there were any references used `this` in `isCardAvailableTarget` or `isAvailableCard`, the function would be dumped with a reference error